### ISSUE TITLE
Consider replacing Pandas

### ISSUE THREAD

**Issue by Innixma** [awslabs/autogluon#1368]
Pandas operations can dictate a large portion of the latency overhead in online inference.

Consider investigating alternative packages:

1. [Modin](https://github.com/modin-project/modin)
2. [Polars](https://pola-rs.github.io/polars-book/user-guide/index.html)

**Comment by willsmithorg** [comment]
Going with a less mainstream package can cause technical debt down the line, and adds to the long list of dependencies.  I would personally stick with Pandas - it's going to be well maintained forever - and try to fix the latency on a case-by-case basis.

**Comment by Innixma** [comment]
> Going with a less mainstream package can cause technical debt down the line, and adds to the long list of dependencies. I would personally stick with Pandas - it's going to be well maintained forever - and try to fix the latency on a case-by-case basis.

I agree. This issue is primarily targeted at extreme low-latency environments (such as requiring <30 ms end-to-end) where pandas overhead starts to be painful. In general I want to avoid switching away from pandas due to a mountain of headaches that could cause, and I would consider this issue low priority compared to other lower hanging fruit.

**Comment by arturdaraujo** [comment]
I think that's worth reconsider using polars. The package is much more stable and good when dealing with larger datasets.

**Comment by jmakov** [comment]
Or https://github.com/narwhals-dev/narwhals (extensible compatibility layer between Polars, pandas, Modin, and cuDF )

**Comment by Innixma** [comment]
One of the bigger blockers to adopting these other options is that they don't have full API support and can differ in API to do complex actions compared to pandas. I've confirmed that polars is indeed faster, but takes substantial effort to get that speed and won't be able to do everything we want it to do.

The most promising use-case is for 1 GB+ sized dataset feature preprocessing and transformation, but we don't see that often enough to have invested in custom solutions yet, and the most time consuming of those (Text processing and Datetime processing) aren't supported in an identical way from what I've seen (No .to_datetime() equivalent in polars, no equivalent alternatives to sklearn's CountVectorizer).

**Comment by jmakov** [comment]
Guess one could ask the guys at https://github.com/TracecatHQ/functime how they managed to do it.

**Comment by mikeman89** [comment]
If you are considering it maybe (depending on how in depth the use is) you can use https://github.com/narwhals-dev/narwhals which is a generic dataframe library that maps over multiple types so this would work with Polars, Pandas, PyArrow etc...

Again i dont know the extent of pandas usage but if its simpler tasks that can be replaced by Narwhals you get the ability to extend to multiple libraries for free.

**Comment by Sdoof** [comment]
Other package that seems worth considering based on benchmark tests below?

[fireducks.pandas](https://fireducks-dev.github.io/docs/benchmarks/)

**Comment by harshitkgupta** [comment]
I see https://modin.readthedocs.io/en/latest/ claims to be drop in replacement for pandas. We should try adding Modin with some flag. eg. use environment variable to enable modin.

**Comment by YuanfengZhang** [comment]
+ Apache-based solutions, including pyarrow and polars, are good at IO, distributed computing and  parallelism.

+ Pandas is now tranferring itself from numpy to pyarrow too, for better data types and computing speed.

+ NVIDIA RAPIDS (including cuML, cuDF) is now working on integrating pandas and polars. They are already usable now.

+ Narwhals does support both arrow- and cuda-based dataframes.

+ Apache arrow and NVIDIA RAPIDS are both widely used and well mantained by big teams.

I think Narwhals is the right way to go. On the other hand, solitary support for customized pandas, such as modin, dask and fireducks, would increase the workload and lose the long-time sustainability. After all, there is no need to bypass Narwhals. They should be integrated in Narwhals first. Even if it get archived someday, autogluon can easily and safely fall back to polars / pyarrow and cuDF by creating a similar and lightweight compatibility layer.

**Comment by shchur** [comment]
Given how small the API surface is for AutoGluon (users only interact with one predictor class and rarely need to chain the operations), I feel that narwhals might be an overkill. If we want to support polars, it would be sufficient to 
1. pick a single internal representation for the DataFrames inside AutoGluon (e.g., always use `polars.DataFrame` or `pandas.DataFrame`).
2. make sure that all predictor methods first convert the inputs to this internal representation, and then convert the outputs back to the format initially provided by the user::

```python
@overload
def predict(self, data: pl.DataFrame, **kwargs) -> pl.DataFrame: ...
@overload
def predict(self, data: pd.DataFrame, **kwargs) -> pd.DataFrame: ...
```

This way we can decouple the predictor API from the AutoGluon internals.

**Comment by MarcoGorelli** [comment]
Thanks for looking into this!

> make sure that all predictor methods first convert the inputs to this internal representation, and then convert the outputs back to the format initially provided by the user:

That's what Plotly used to do, and [plots generally got 3x or even 10x faster](https://plotly.com/blog/chart-smarter-not-harder-universal-dataframe-support/) when they switched to supporting each dataframe natively

It can be quite expensive to convert between polars and pandas for certain data types (like strings) so i think many users would consider this to be a missed opportunity. It also forces all Polars users to take on pandas (and probably PyArrow) as dependencies, increasing the risk of [dependency hell](https://en.wikipedia.org/wiki/Dependency_hell)

**Comment by YuanfengZhang** [comment]
From the perspective of computational efficiency, currently polars is much better than pandas at feature preprocessing, dtype conversion, metric calucaltion etc throught its native support for multiprocessing on CPU and GPU, as well as lazy evaluation.
+ For pytorch-based models, pl.DataFrame has to_torch() method.
+ LightGBM now supports pl.DataFrame as input:
https://github.com/microsoft/LightGBM/issues/6204
https://github.com/microsoft/LightGBM/pull/6910
+ CatBoost now supports pl.DataFrame as input:
https://github.com/dmlc/xgboost/pull/11116
+ Scikit-learn now supports pl.DataFrame:
https://github.com/scikit-learn/scikit-learn/issues/28341

Pandas is only need in [eda/visualization](https://github.com/autogluon/autogluon/blob/d3aa066169eba691d10d21e406f140f13ec52d90/eda/src/autogluon/eda/visualization/model.py#L5) because seaborn and matplotlib do not support polars. However, to_pandas() for visualization is acceptable. Except for this, After a rough scaning, I think all pandas and numpy lines in autogluon repo can be safely replaced with polars and improved.

For example:
https://github.com/autogluon/autogluon/blob/d3aa066169eba691d10d21e406f140f13ec52d90/common/src/autogluon/common/utils/utils.py#L194-L204

This function hashes all the rows in the dataframe and can be modified as:
```python
import polars as pl


def hash_polars_df(df: pl.DataFrame | None) -> str:
    """Compute a hash string for a polars DataFrame."""
    if df is not None:
        hashed_rows = df.hash_rows()
    else:
        hashed_rows = "0".encode("utf-8")
    return md5(hashed_rows).hexdigest()
```
No full copy, no index issue, blazingly fast and resource-saving.
If you want to hash the whole dataframe, just use write_ipc() method:
```python
from io import BytesIO


def hash_whole_df(df: pl.DataFrame | None) -> str:
    """Compute a hash string for a polars DataFrame."""
    if df is not None:
        with BytesIO() as buffer:
            df.write_ipc(buffer, compression=None)
        return md5(buffer.getvalue()).hexdigest()
    else:
        return md5(b"0").hexdigest()
```

Another instance: metrics like RMSE now are calucalted using scikit-learn based on numpy:
https://github.com/autogluon/autogluon/blob/d3aa066169eba691d10d21e406f140f13ec52d90/tabular/src/autogluon/tabular/models/mitra/_internal/core/prediction_metrics.py#L81

which is very, very slow for large datasets. While in polars, calculating metrics like RMSE can be very fast:
```python
import polars as pl


def root_mean_squared_error(df: pl.DataFrame, y_pred: str, y_true: str) -> float:
    return (df.lazy()
              .with_columns((pl.col(y_pred) - pl.col(y_true))
                            .pow(2).mean().sqrt())
              .collect()  # if `pip install polars[gpu]`, we can use .collect('gpu') here
              .item())
```

We can also use [3rd-party solutions](https://github.com/abstractqqq/polars_ds_extension) or write [our own polars plugins](https://docs.pola.rs/user-guide/plugins/#various)

Besides, I don't think it will be troublesome for dependency, causing I've been using polars[pandas,pyarrow] and autogluon for months. Polars is a very lightweight lib.

In conclusion, I suggest migration from numpy+pandas to arrow+polars to meet the needs of large dataset and quick inference.

**Comment by YuanfengZhang** [comment]
☝️🤓The data types, type annotation and code annotation of polars are also better than those of pandas. I gotta say pandas and numpy are always disastrous for development, according to countless red and yellow warnings pointed out by mypy, pyright and alike tools. Maintainability is surely important.

**Comment by MarcoGorelli** [comment]
> Pandas is only need in [eda/visualization](https://github.com/autogluon/autogluon/blob/d3aa066169eba691d10d21e406f140f13ec52d90/eda/src/autogluon/eda/visualization/model.py#L5) because seaborn and matplotlib do not support polars

Minor note: it's true that Seaborn doesn't (or rather, it converts everything to pandas anyway 😩 ), but [Matplotlib can accept Polars Series without going via NumPy](https://docs.pola.rs/user-guide/misc/visualization/#matplotlib)

**Comment by YuanfengZhang** [comment]
> One of the bigger blockers to adopting these other options is that they don't have full API support and can differ in API to do complex actions compared to pandas. I've confirmed that polars is indeed faster, but takes substantial effort to get that speed and won't be able to do everything we want it to do.
> 
> The most promising use-case is for 1 GB+ sized dataset feature preprocessing and transformation, but we don't see that often enough to have invested in custom solutions yet, and the most time consuming of those (Text processing and Datetime processing) aren't supported in an identical way from what I've seen (No .to_datetime() equivalent in polars, no equivalent alternatives to sklearn's CountVectorizer).

I'm not familiar with CountVectorizer but scikit-learn now supports polars DataFrame. Also, `pl.Series.str.to_time()` [here](https://docs.pola.rs/api/python/stable/reference/series/api/polars.Series.str.to_time.html#polars.Series.str.to_time) is a flexiable version of `pandas.to_datetime`. With methods under [pd.Series.Temporal](https://docs.pola.rs/api/python/stable/reference/series/temporal.html) and [pd.Expression.Temporal](https://docs.pola.rs/api/python/stable/reference/expressions/temporal.html), the timeseries processing is not that difficult. Maybe we could make a move now?

**Comment by Sdoof** [comment]
Hi Gang, given that there seem to be a genuine interest I was wondering what were the nest steps. @YuanfengZhang has suggested what seems a viable path forward,

**Comment by nelsonrogers-flowline** [comment]
Any news on this?
